Fix for issue #559 - Migrate from Yeller to Sentry

kahunamoore opened the pull request:
First pass at converting from Yeller to Sentry error tracking service. Uses raven-clj as the Sentry client.

danielcompton commented on src/clojars/errors.clj:
```diff
@@ -1,21 +1,28 @@
 (defprotocol ErrorReporter
   (-report-error [reporter e extra id]))
 
-(extend-protocol ErrorReporter
```
Might pay to leave the Yeller reporter in for a bit longer to check if they are both reporting on the same issues?

danielcompton commented on src/clojars/errors.clj:
```diff
@@ -43,8 +50,20 @@
+                         (interfaces/stacktrace e ["clojars"])
+                         (interfaces/http request alter-fn))))
+
+; Note: I did not use raven-ring because this exisitng logic wraps ring exceptions
```
Can probably remove this comment?

danielcompton commented on src/clojars/main.clj:
```diff
@@ -28,16 +28,15 @@
-                    (:queue system)
-                    (:search system)
-                    (:storage system))))
+    (println "clojars-web: enabling raven-clj client dsn:project-id:" (:project-id (raven-clj/parse-dsn (:raven-dsn config))))
```
Would prefer if this was called `:sentry-dsn` I think

danielcompton replied:
This might be clearer with a `->`?

danielcompton commented on src/clojars/errors.clj:
```diff
@@ -43,8 +50,20 @@
+  (raven-clj/capture (:dsn reporter)
+                     (-> {:message "Ring reported an error"}
+                         (interfaces/stacktrace e ["clojars"])
+                         (interfaces/http request alter-fn))))
```
I think it would still be good to remove information from the request on our side, as it prevents us having to trust Sentry to store/filter sensitive info.

danielcompton commented on src/clojars/config.clj:
```diff
@@ -71,8 +70,7 @@
    ["REPO" :repo]
-   ["YELLER_ENV" :yeller-environment]
-   ["YELLER_TOKEN" :yeller-token]
+   ["RAVEN_DSN" :raven-dsn]
```
SENTRY_DSN

danielcompton commented on dev-resources/config.clj:
```diff
@@ -10,4 +10,5 @@
  :mail {:hostname "localhost"
         :from "noreply@clojars.org"
         :ssl false}
+ ;; :raven-dsn "https://b70a31b3510c4cf793964a185cfe1fd0:b7d80b520139450f903720eb7991bf3d@sentry.example.com/1"
```
Do we need this here?

kahunamoore commented:
Should I amend this pull request or close it and create a new one? I've addressed most of the issues you mentioned with the exception of reporting to both sentry and yeller. The handling of extra info is different for both so I would have to duplicate a bunch of logic which would complicate the code quite a bit. The reporting to Sentry seems to be working quite well. Is reporting to both necessary? I'll do it if you really want it, I was just trying to keep the code as simple as possible.

kahunamoore commented:
You can see my changes on this branch:

https://github.com/coopsource/clojars-web/tree/yeller-to-sentry

kahunamoore commented:
BTW - I thought I squashed all my commits into one before pushing but it doesn't look like it here on github... not exactly sure what happened. Sorry about that. If you have suggestions on how to squash them post-facto I'll do whatever is necessary. Let me know!

kahunamoore commented:
Sorry for the comment noise... I think I need to push some changes to the clojars server deploy in order to add the SENTRY_DSN environment variable. How should any changes to:

https://github.com/clojars/clojars-server-config

... be coordinated with the pull request for clojars-web?

The DSN contains some sensitive information so I don't know how that is handled given that the repo is public. You can get the DSN from Sentry itself or I can send it to you OOB.

I'm going to take a look at updating clojars-server-config and test those changes on my server.

tobias commented:
Thanks for all your work on this; it looks good to me. We can squash the commits when we merge the PR, so don't worry about doing that. For the DSN, we store the private data elsewhere, so we can update that ourselves before the next deploy.

I think the last step is to have you rebase against master one more time, since I merged https://github.com/clojars/clojars-web/pull/588, which moved us to `juxt/aero` for configuration, and eliminated the use of env vars. All you'll need to do is update your changes to `config.clj` so it merges cleanly, and I'll take care of updating the readme instructions and the server-config repo to handle the DSN. 

Thanks!

tobias commented:
@kahunamoore - actually, I'll fix this merge for you - I want to make some extensive changes to how we handle config (#590) which will create a bit more work for you.

tobias commented:
Merged via: https://github.com/clojars/clojars-web/commit/2e7254b4fe214ed8ca50e12b4bc95844758af3fc
